File: overlays/src/lib/nowPlaying.ts
// ---------------------------------------------------------------------------
// Now-playing view-model mapper
//
// The boundary between the feed (NowPlayingDto wire shape) and the headless
// ui/footer/NowPlaying widget. ui/ never imports feed DTOs; this is where the
// ambient Spotify state is flattened into the props NowPlaying renders.
//
// A "stopped" status (player gone/idle, empty strings) maps to null so the
// widget falls back to its placeholder. "playing" vs "paused" only differ by
// whether the disc/EQ animations run, so we collapse them into `isPlaying`.
// ---------------------------------------------------------------------------

import type { NowPlayingDto } from "../feed";
import type { NowPlayingTrack } from "../ui/footer/NowPlaying";

export function toNowPlaying(dto: NowPlayingDto): NowPlayingTrack | null {
  if (dto.status === "stopped") return null;

  return {
    title: dto.title,
    artist: dto.artist,
    artUrl: dto.artUrl,
    isPlaying: dto.status === "playing",
  };
}

File: src/presentation/http/resources/now_playing.rs
//! Now-playing feed DTO — the ambient Spotify track state.
//!
//! now-playing is ambient *state* (latest value on a watch channel), not a
//! logged event — but it rides the same Overlay Feed so the Coworking Overlay's
//! Now Playing widget can render it. `status` mirrors the domain
//! [`PlaybackStatus`]; the React side falls back to the placeholder on
//! `"stopped"`. `artUrl` is the real `mpris:artUrl` album cover when present.

use serde::Serialize;

use crate::domain::{NowPlaying, PlaybackStatus};

/// The currently-playing Spotify track as the React side consumes it.
#[derive(Debug, Clone, Serialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct NowPlayingDto {
    pub title: String,
    pub artist: String,
    pub album: String,
    pub art_url: Option<String>,
    pub status: &'static str,
}

impl NowPlayingDto {
    /// A "cleared" now-playing DTO — emitted when playback stops or no player is
    /// present. Empty strings + no art + `"stopped"` status drive the Now Playing
    /// widget back to its placeholder.
    pub fn cleared() -> Self {
        NowPlayingDto {
            title: String::new(),
            artist: String::new(),
            album: String::new(),
            art_url: None,
            status: "stopped",
        }
    }
}

impl From<&NowPlaying> for NowPlayingDto {
    fn from(now_playing: &NowPlaying) -> Self {
        let status = match now_playing.status {
            PlaybackStatus::Playing => "playing",
            PlaybackStatus::Paused => "paused",
            PlaybackStatus::Stopped => "stopped",
        };
        NowPlayingDto {
            title: now_playing.title.clone(),
            artist: now_playing.artist.clone(),
            album: now_playing.album.clone(),
            art_url: now_playing.art_url.clone(),
            status,
        }
    }
}
